File: code/python/03-agent-loop/plan_execute_agent.py
"""Plan-and-Execute agent: separate planning from execution.

Implements the three-phase pattern described in:
docs/02-the-agent-loop/03-planning-strategies.md

Phase 1 PLAN:   LLM generates a structured list of steps.
Phase 2 EXECUTE: Steps run in dependency order; independent steps run in parallel.
Phase 3 SYNTHESIZE: LLM combines all results into a final answer.

Run:
    python plan_execute_agent.py
"""

from __future__ import annotations

import json
import logging
import os
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any, Optional

from openai import OpenAI
from pydantic import ValidationError

from plan_schema import AgentPlan, PlanStep, StepResult

logger = logging.getLogger(__name__)

MAX_STEPS = 20


# ---------------------------------------------------------------------------
# Mock tools (same data as tools.py so the agent is self-contained)
# ---------------------------------------------------------------------------

_WEATHER_MOCK: dict[str, dict] = {
    "Apple":     {"price_usd": 192.35, "change_percent":  1.2, "currency": "USD"},  # kept for compat
    "Shanghai":  {"temperature_c": 22, "condition": "light rain",   "humidity_percent": 85, "wind_kph": 15},
    "London":    {"temperature_c": 14, "condition": "overcast",     "humidity_percent": 78, "wind_kph": 20},
    "New York":  {"temperature_c": 18, "condition": "partly cloudy","humidity_percent": 60, "wind_kph": 25},
    "Tokyo":     {"temperature_c": 22, "condition": "light rain",   "humidity_percent": 85, "wind_kph": 15},
    "Paris":     {"temperature_c": 16, "condition": "sunny",        "humidity_percent": 55, "wind_kph": 12},
    "Sydney":    {"temperature_c": 28, "condition": "clear",        "humidity_percent": 45, "wind_kph": 18},
}

_STOCK_MOCK: dict[str, dict] = {
    "AAPL":  {"price_usd": 192.35, "change_percent":  1.2, "currency": "USD", "weekly_change_percent": 3.1},
    "GOOGL": {"price_usd": 171.80, "change_percent": -0.5, "currency": "USD", "weekly_change_percent": 1.5},
    "MSFT":  {"price_usd": 415.10, "change_percent":  0.8, "currency": "USD", "weekly_change_percent": 2.4},
    "TSLA":  {"price_usd": 175.20, "change_percent": -2.3, "currency": "USD", "weekly_change_percent": -4.1},
    "AMZN":  {"price_usd": 188.40, "change_percent":  0.3, "currency": "USD", "weekly_change_percent": 0.8},
}

_NEWS_MOCK: dict[str, list[dict]] = {
    "AAPL": [
        {"headline": "Apple unveils M4 Ultra chip with record AI performance", "sentiment": "positive"},
        {"headline": "iPhone 17 pre-orders exceed expectations", "sentiment": "positive"},
    ],
    "MSFT": [
        {"headline": "Microsoft Azure revenue grows 33% year-over-year", "sentiment": "positive"},
        {"headline": "Copilot+ PC sales drive record Surface quarter", "sentiment": "positive"},
    ],
    "TSLA": [
        {"headline": "Tesla delays Robotaxi launch to Q3", "sentiment": "negative"},
        {"headline": "Tesla energy storage deployments hit record", "sentiment": "positive"},
    ],
}


def _get_weather(city: str) -> dict:
    key = city.split(",")[0].strip()
    data = _WEATHER_MOCK.get(key, {"temperature_c": 20, "condition": "clear", "humidity_percent": 55, "wind_kph": 10})
    return {"city": city, **data}


def _get_stock_price(ticker: str) -> dict:
    upper = ticker.upper()
    data = _STOCK_MOCK.get(upper, {"price_usd": 100.0, "change_percent": 0.0, "currency": "USD", "weekly_change_percent": 0.0})
    return {"ticker": upper, "market_status": "open", **data}


def _search_news(query: str) -> dict:
    ticker = query.upper().split()[0]
    articles = _NEWS_MOCK.get(ticker, [
        {"headline": f"No major news found for '{query}'", "sentiment": "neutral"}
    ])
    return {"query": query, "articles": articles, "count": len(articles)}


_TOOL_DISPATCH: dict[str, Any] = {
    "get_weather":     _get_weather,
    "get_stock_price": _get_stock_price,
    "search_news":     _search_news,
}

TOOLS: list[dict] = [
    {
        "type": "function",
        "function": {
            "name": "get_weather",
            "description": (
                "Get current weather conditions for a city. "
                "Returns temperature (C), humidity, wind speed, and condition."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "city": {
                        "type": "string",
                        "description": "City name, e.g. 'Tokyo' or 'New York'.",
                    }
                },
                "required": ["city"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "get_stock_price",
            "description": (
                "Get current stock price and weekly percentage change. "
                "Returns price_usd, change_percent, and weekly_change_percent."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "ticker": {
                        "type": "string",
                        "description": "Stock ticker symbol in uppercase, e.g. 'AAPL'.",
                    }
                },
                "required": ["ticker"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "search_news",
            "description": (
                "Search for recent news headlines about a company or topic. "
                "Returns a list of headlines with sentiment labels."
            ),
            "parameters": {
                "type": "object",
                "properties": {
                    "query": {
                        "type": "string",
                        "description": "Company name or ticker, e.g. 'AAPL' or 'Apple'.",
                    }
                },
                "required": ["query"],
            },
        },
    },
]


# ---------------------------------------------------------------------------
# Planner and synthesizer system prompts
# ---------------------------------------------------------------------------

_PLANNER_SYSTEM = """\
You are a planning assistant. Break the user's request into concrete, sequential steps.

Rules:
- Each step must have a unique sequential step_number starting at 1.
- tool_name must be one of: get_weather, get_stock_price, search_news — or null for reasoning/synthesis steps.
- tool_params must contain the exact arguments to pass to the tool (or null for reasoning steps).
- depends_on lists the step_numbers this step must wait for.
- expected_output describes what a successful result looks like (at least 15 characters).
- Keep the plan to at most 10 steps. Consolidate where possible.

Output ONLY valid JSON matching this schema:
{
  "user_question": "<the user's question>",
  "steps": [
    {
      "step_number": 1,
      "description": "...",
      "tool_name": "get_stock_price" | null,
      "tool_params": {"ticker": "AAPL"} | null,
      "depends_on": [],
      "expected_output": "AAPL price and weekly percentage change"
    }
  ],
  "estimated_tool_calls": <int>
}"""

_EXECUTOR_SYSTEM = """\
You are an executor. Complete the given reasoning step using the provided context.
Be concise and factual. Return only the answer to this specific step."""

_SYNTHESIZER_SYSTEM = """\
You are a synthesizer. Given the original question and all execution results,
write a complete, well-structured answer. Cite specific numbers and data from
the results. Use markdown formatting where it improves readability."""


# ---------------------------------------------------------------------------
# PlanAndExecuteAgent
# ---------------------------------------------------------------------------


class PlanAndExecuteAgent:
    """Three-phase agent: Plan → Execute → Synthesize.

    Args:
        tools:      List of OpenAI tool definition dicts (the function schemas).
        model:      OpenAI model name for all LLM calls.
        max_steps:  Safety limit on plan length.
    """

    def __init__(
        self,
        tools: list[dict] = TOOLS,
        model: str = "gpt-4o",
        max_steps: int = MAX_STEPS,
    ) -> None:
        self.tools = tools
        self.model = model
        self.max_steps = max_steps
        self._client: Optional[OpenAI] = None

    @property
    def client(self) -> OpenAI:
        if self._client is None:
            self._client = OpenAI(api_key=os.environ.get("OPENAI_API_KEY", ""))
        return self._client

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def run(self, user_input: str) -> dict[str, Any]:
        """Run the full Plan → Execute → Synthesize cycle.

        Args:
            user_input: The user's question or request.

        Returns:
            A dict with keys:
                ``plan``    — list of PlanStep dicts
                ``results`` — list of StepResult dicts
                ``answer``  — final synthesised answer string
        """
        logger.info("[PlanAndExecute] Planning for: %s", user_input)
        plan = self._generate_plan(user_input)
        logger.info("[PlanAndExecute] Plan has %d steps", len(plan))

        results = self._execute_plan(plan)

        successes = sum(1 for r in results if r.success)
        logger.info("[PlanAndExecute] Executed %d/%d steps successfully", successes, len(results))

        answer = self._synthesize(user_input, plan, results)

        return {
            "plan": [s.model_dump() for s in plan],
            "results": [r.model_dump() for r in results],
            "answer": answer,
        }

    # ------------------------------------------------------------------
    # Phase 1: Plan
    # ------------------------------------------------------------------

    def _generate_plan(self, user_input: str) -> list[PlanStep]:
        """Ask the LLM to produce a structured JSON plan.

        Falls back to a single-step plan if the LLM output cannot be parsed.
        """
        response = self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": _PLANNER_SYSTEM},
                {"role": "user", "content": user_input},
            ],
            response_format={"type": "json_object"},
            temperature=0,
        )
        raw = response.choices[0].message.content or "{}"
        try:
            data = json.loads(raw)
            plan_obj = AgentPlan.model_validate(data)
            steps = plan_obj.steps[: self.max_steps]
        except (json.JSONDecodeError, ValidationError) as exc:
            logger.warning("[PlanAndExecute] Plan parsing failed (%s); using fallback", exc)
            steps = [
                PlanStep(
                    step_number=1,
                    description="Answer the user's question directly using available tools.",
                    tool_name=None,
                    tool_params=None,
                    depends_on=[],
                    expected_output="A complete answer to the user's question",
                )
            ]
        return steps

    # ------------------------------------------------------------------
    # Phase 2: Execute
    # ------------------------------------------------------------------

    def _execute_plan(self, steps: list[PlanStep]) -> list[StepResult]:
        """Execute plan steps, respecting dependencies and running independent
        steps in parallel.
        """
        results: dict[int, StepResult] = {}

        # Topological execution: keep finding steps whose dependencies are met
        remaining = list(steps)
        max_rounds = len(steps) + 1  # safety

        for _ in range(max_rounds):
            if not remaining:
                break
            ready = [s for s in remaining if all(dep in results for dep in s.depends_on)]
            if not ready:
                # Unresolvable — mark everything left as failed
                for s in remaining:
                    results[s.step_number] = StepResult(
                        step_number=s.step_number,
                        success=False,
                        error="Could not execute: unresolved dependencies",
                    )
                break

            # Execute all ready steps in parallel
            with ThreadPoolExecutor(max_workers=min(len(ready), 8)) as pool:
                futures = {
                    pool.submit(self._execute_step, step, results): step
                    for step in ready
                }
                for future in as_completed(futures):
                    step = futures[future]
                    result = future.result()
                    results[step.step_number] = result

            for s in ready:
                remaining.remove(s)

        return [results[s.step_number] for s in steps]

    def _execute_step(
        self, step: PlanStep, prior_results: dict[int, StepResult]
    ) -> StepResult:
        """Execute a single plan step: call a tool or ask the LLM to reason."""
        start = time.monotonic()
        try:
            if step.tool_name:
                data = self._call_tool(step.tool_name, step.tool_params or {})
            else:
                data = self._reason(step, prior_results)

            duration_ms = int((time.monotonic() - start) * 1000)
            return StepResult(
                step_number=step.step_number,
                success=True,
                data=data,
                duration_ms=duration_ms,
            )
        except Exception as exc:  # noqa: BLE001
            duration_ms = int((time.monotonic() - start) * 1000)
            logger.warning("[PlanAndExecute] Step %d failed: %s", step.step_number, exc)
            return StepResult(
                step_number=step.step_number,
                success=False,
                error=str(exc),
                duration_ms=duration_ms,
            )

    def _call_tool(self, tool_name: str, params: dict[str, Any]) -> dict[str, Any]:
        """Dispatch a tool call to the local mock implementations."""
        fn = _TOOL_DISPATCH.get(tool_name)
        if fn is None:
            raise ValueError(f"Unknown tool: '{tool_name}'")
        result = fn(**params)
        if not isinstance(result, dict):
            result = {"result": result}
        return result

    def _reason(
        self, step: PlanStep, prior_results: dict[int, StepResult]
    ) -> dict[str, Any]:
        """Ask the LLM to reason about a non-tool step, given prior results."""
        context_parts = []
        for dep_num in step.depends_on:
            dep_result = prior_results.get(dep_num)
            if dep_result and dep_result.success:
                context_parts.append(
                    f"Step {dep_num} result:\n{json.dumps(dep_result.data, indent=2)}"
                )

        user_message = step.description
        if context_parts:
            user_message += "\n\nContext from previous steps:\n" + "\n\n".join(context_parts)

        response = self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": _EXECUTOR_SYSTEM},
                {"role": "user", "content": user_message},
            ],
            temperature=0.3,
        )
        return {"reasoning": response.choices[0].message.content or ""}

    # ------------------------------------------------------------------
    # Phase 3: Synthesize
    # ------------------------------------------------------------------

    def _synthesize(
        self,
        question: str,
        steps: list[PlanStep],
        results: list[StepResult],
    ) -> str:
        """Combine all step results into a final answer."""
        lines = [f"Original question: {question}\n"]
        for step, result in zip(steps, results):
            status = "✓" if result.success else "✗"
            lines.append(f"Step {step.step_number} [{status}]: {step.description}")
            if result.success and result.data:
                lines.append(f"Result: {json.dumps(result.data)}")
            elif result.error:
                lines.append(f"Error: {result.error}")
            lines.append("")

        context = "\n".join(lines)

        response = self.client.chat.completions.create(
            model=self.model,
            messages=[
                {"role": "system", "content": _SYNTHESIZER_SYSTEM},
                {"role": "user", "content": context},
            ],
            temperature=0.5,
        )
        return response.choices[0].message.content or ""


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def main() -> None:
    logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")

    agent = PlanAndExecuteAgent()
    query = "Compare Apple and Microsoft stock performance this week"

    print(f"\n{'=' * 60}")
    print(f"Query: {query}")
    print("=" * 60)

    output = agent.run(query)

    print("\n--- Plan ---")
    for step in output["plan"]:
        deps = f" (depends on {step['depends_on']})" if step["depends_on"] else ""
        tool = f" [{step['tool_name']}]" if step["tool_name"] else " [reasoning]"
        print(f"  {step['step_number']}. {step['description']}{tool}{deps}")

    print("\n--- Results ---")
    for result in output["results"]:
        status = "✓" if result["success"] else "✗"
        duration = result.get("duration_ms", 0)
        print(f"  Step {result['step_number']} [{status}] ({duration}ms)")
        if result["success"] and result.get("data"):
            print(f"    {json.dumps(result['data'])}")
        elif result.get("error"):
            print(f"    ERROR: {result['error']}")

    print("\n--- Final Answer ---")
    print(output["answer"])


if __name__ == "__main__":
    main()

File: code/nodejs/03-agent-loop/plan_execute_agent.ts
/**
 * Plan-and-Execute agent in TypeScript.
 *
 * Three-phase pattern:
 *   Phase 1 PLAN:      LLM generates a structured JSON plan.
 *   Phase 2 EXECUTE:   Steps run in dependency order; independent steps in parallel.
 *   Phase 3 SYNTHESIZE: LLM combines results into a final answer.
 *
 * Structurally equivalent to code/python/03-agent-loop/plan_execute_agent.py.
 *
 * Run:  npx tsx plan_execute_agent.ts
 * See:  docs/02-the-agent-loop/03-planning-strategies.md
 */

import OpenAI from "openai";
import { z } from "zod";

// ---------------------------------------------------------------------------
// Zod schemas (equivalent to plan_schema.py)
// ---------------------------------------------------------------------------

export const PlanStepSchema = z.object({
  step_number: z.number().int().min(1),
  description: z.string().min(10).max(500),
  tool_name: z.enum(["get_weather", "get_stock_price", "search_news"]).nullable(),
  tool_params: z.record(z.unknown()).nullable(),
  depends_on: z.array(z.number().int()).default([]),
  expected_output: z.string().min(5),
});

export const AgentPlanSchema = z.object({
  user_question: z.string(),
  steps: z.array(PlanStepSchema).min(1),
  estimated_tool_calls: z.number().int().default(0),
});

export const StepResultSchema = z.object({
  step_number: z.number().int().min(1),
  success: z.boolean(),
  data: z.record(z.unknown()).nullable().default(null),
  error: z.string().nullable().default(null),
  duration_ms: z.number().int().min(0).default(0),
});

export type PlanStep = z.infer<typeof PlanStepSchema>;
export type AgentPlan = z.infer<typeof AgentPlanSchema>;
export type StepResult = z.infer<typeof StepResultSchema>;

// ---------------------------------------------------------------------------
// Mock tool implementations
// ---------------------------------------------------------------------------

const WEATHER_MOCK: Record<string, Record<string, unknown>> = {
  Tokyo:     { temperature_c: 22, condition: "light rain",    humidity_percent: 85, wind_kph: 15 },
  London:    { temperature_c: 14, condition: "overcast",      humidity_percent: 78, wind_kph: 20 },
  "New York":{ temperature_c: 18, condition: "partly cloudy", humidity_percent: 60, wind_kph: 25 },
  Paris:     { temperature_c: 16, condition: "sunny",         humidity_percent: 55, wind_kph: 12 },
  Sydney:    { temperature_c: 28, condition: "clear",         humidity_percent: 45, wind_kph: 18 },
};

const STOCK_MOCK: Record<string, Record<string, unknown>> = {
  AAPL:  { price_usd: 192.35, change_percent:  1.2, currency: "USD", weekly_change_percent:  3.1 },
  GOOGL: { price_usd: 171.80, change_percent: -0.5, currency: "USD", weekly_change_percent:  1.5 },
  MSFT:  { price_usd: 415.10, change_percent:  0.8, currency: "USD", weekly_change_percent:  2.4 },
  TSLA:  { price_usd: 175.20, change_percent: -2.3, currency: "USD", weekly_change_percent: -4.1 },
  AMZN:  { price_usd: 188.40, change_percent:  0.3, currency: "USD", weekly_change_percent:  0.8 },
};

const NEWS_MOCK: Record<string, Array<{ headline: string; sentiment: string }>> = {
  AAPL: [
    { headline: "Apple unveils M4 Ultra chip with record AI performance", sentiment: "positive" },
    { headline: "iPhone 17 pre-orders exceed expectations",                sentiment: "positive" },
  ],
  MSFT: [
    { headline: "Microsoft Azure revenue grows 33% year-over-year",       sentiment: "positive" },
    { headline: "Copilot+ PC sales drive record Surface quarter",          sentiment: "positive" },
  ],
};

function getWeather(city: string): Record<string, unknown> {
  const key = city.split(",")[0].trim();
  return { city, ...(WEATHER_MOCK[key] ?? { temperature_c: 20, condition: "clear" }) };
}

function getStockPrice(ticker: string): Record<string, unknown> {
  const upper = ticker.toUpperCase();
  return { ticker: upper, market_status: "open", ...(STOCK_MOCK[upper] ?? { price_usd: 100, change_percent: 0 }) };
}

function searchNews(query: string): Record<string, unknown> {
  const ticker = query.toUpperCase().split(" ")[0];
  const articles = NEWS_MOCK[ticker] ?? [{ headline: `No major news for '${query}'`, sentiment: "neutral" }];
  return { query, articles, count: articles.length };
}

type ToolFn = (params: Record<string, unknown>) => Record<string, unknown>;

const TOOL_DISPATCH: Record<string, ToolFn> = {
  get_weather:     (p) => getWeather(p["city"] as string),
  get_stock_price: (p) => getStockPrice(p["ticker"] as string),
  search_news:     (p) => searchNews(p["query"] as string),
};

// ---------------------------------------------------------------------------
// Planner / synthesizer prompts
// ---------------------------------------------------------------------------

const PLANNER_SYSTEM = `You are a planning assistant. Break the user's request into concrete sequential steps.

Rules:
- step_number must start at 1 and increment by 1.
- tool_name must be one of: get_weather, get_stock_price, search_news — or null for reasoning/synthesis steps.
- tool_params must contain the exact arguments (or null for reasoning steps).
- depends_on lists the step_numbers this step must wait for.
- expected_output describes the result (at least 15 chars).
- Keep the plan to at most 10 steps.

Output ONLY valid JSON:
{
  "user_question": "<question>",
  "steps": [{"step_number":1,"description":"...","tool_name":"get_stock_price","tool_params":{"ticker":"AAPL"},"depends_on":[],"expected_output":"AAPL price and weekly change"}],
  "estimated_tool_calls": 1
}`;

const SYNTHESIZER_SYSTEM = `You are a synthesizer. Given the original question and all execution results,
write a complete, well-structured answer. Cite specific numbers and data from the results.
Use markdown formatting where it improves readability.`;

// ---------------------------------------------------------------------------
// PlanAndExecuteAgent
// ---------------------------------------------------------------------------

export class PlanAndExecuteAgent {
  private readonly model: string;
  private readonly maxSteps: number;
  private readonly client: OpenAI;

  constructor(
    model = "gpt-4o",
    maxSteps = 20,
    apiKey = process.env["OPENAI_API_KEY"] ?? "",
  ) {
    this.model = model;
    this.maxSteps = maxSteps;
    this.client = new OpenAI({ apiKey });
  }

  /** Run the full Plan → Execute → Synthesize cycle. */
  async run(userInput: string): Promise<{
    plan: PlanStep[];
    results: StepResult[];
    answer: string;
  }> {
    const plan = await this.generatePlan(userInput);
    const results = await this.executePlan(plan);
    const answer = await this.synthesize(userInput, plan, results);
    return { plan, results, answer };
  }

  // ------------------------------------------------------------------
  // Phase 1: Plan
  // ------------------------------------------------------------------

  private async generatePlan(userInput: string): Promise<PlanStep[]> {
    const response = await this.client.chat.completions.create({
      model: this.model,
      messages: [
        { role: "system", content: PLANNER_SYSTEM },
        { role: "user",   content: userInput },
      ],
      response_format: { type: "json_object" },
      temperature: 0,
    });

    const raw = response.choices[0]?.message.content ?? "{}";
    try {
      const parsed = JSON.parse(raw) as unknown;
      const planObj = AgentPlanSchema.parse(parsed);
      return planObj.steps.slice(0, this.maxSteps);
    } catch {
      // Fallback to single reasoning step
      return [
        {
          step_number:     1,
          description:     "Answer the user's question using available tools.",
          tool_name:       null,
          tool_params:     null,
          depends_on:      [],
          expected_output: "A complete answer to the user's question",
        },
      ];
    }
  }

  // ------------------------------------------------------------------
  // Phase 2: Execute
  // ------------------------------------------------------------------

  private async executePlan(steps: PlanStep[]): Promise<StepResult[]> {
    const results = new Map<number, StepResult>();
    let remaining = [...steps];
    const maxRounds = steps.length + 1;

    for (let round = 0; round < maxRounds && remaining.length > 0; round++) {
      const ready = remaining.filter((s) =>
        s.depends_on.every((dep) => results.has(dep)),
      );
      if (ready.length === 0) {
        // Unresolvable dependencies
        for (const s of remaining) {
          results.set(s.step_number, {
            step_number: s.step_number,
            success:     false,
            data:        null,
            error:       "Unresolved dependencies",
            duration_ms: 0,
          });
        }
        break;
      }

      // Execute ready steps in parallel
      const settled = await Promise.allSettled(
        ready.map((step) => this.executeStep(step, results)),
      );

      for (let i = 0; i < ready.length; i++) {
        const step = ready[i]!;
        const outcome = settled[i]!;
        if (outcome.status === "fulfilled") {
          results.set(step.step_number, outcome.value);
        } else {
          results.set(step.step_number, {
            step_number: step.step_number,
            success:     false,
            data:        null,
            error:       String(outcome.reason),
            duration_ms: 0,
          });
        }
        remaining = remaining.filter((s) => s.step_number !== step.step_number);
      }
    }

    return steps.map((s) => results.get(s.step_number)!);
  }

  private async executeStep(
    step: PlanStep,
    priorResults: Map<number, StepResult>,
  ): Promise<StepResult> {
    const start = Date.now();
    try {
      let data: Record<string, unknown>;
      if (step.tool_name !== null) {
        data = this.callTool(step.tool_name, (step.tool_params as Record<string, unknown>) ?? {});
      } else {
        data = await this.reason(step, priorResults);
      }
      return {
        step_number: step.step_number,
        success:     true,
        data,
        error:       null,
        duration_ms: Date.now() - start,
      };
    } catch (err) {
      return {
        step_number: step.step_number,
        success:     false,
        data:        null,
        error:       String(err),
        duration_ms: Date.now() - start,
      };
    }
  }

  private callTool(
    toolName: string,
    params: Record<string, unknown>,
  ): Record<string, unknown> {
    const fn = TOOL_DISPATCH[toolName];
    if (!fn) throw new Error(`Unknown tool: '${toolName}'`);
    return fn(params);
  }

  private async reason(
    step: PlanStep,
    priorResults: Map<number, StepResult>,
  ): Promise<Record<string, unknown>> {
    const contextParts: string[] = [];
    for (const depNum of step.depends_on) {
      const dep = priorResults.get(depNum);
      if (dep?.success && dep.data) {
        contextParts.push(`Step ${depNum} result:\n${JSON.stringify(dep.data, null, 2)}`);
      }
    }

    const userMessage =
      step.description +
      (contextParts.length > 0
        ? "\n\nContext from previous steps:\n" + contextParts.join("\n\n")
        : "");

    const response = await this.client.chat.completions.create({
      model: this.model,
      messages: [
        { role: "system", content: "You are an executor. Complete the given reasoning step concisely." },
        { role: "user",   content: userMessage },
      ],
      temperature: 0.3,
    });
    return { reasoning: response.choices[0]?.message.content ?? "" };
  }

  // ------------------------------------------------------------------
  // Phase 3: Synthesize
  // ------------------------------------------------------------------

  private async synthesize(
    question: string,
    steps: PlanStep[],
    results: StepResult[],
  ): Promise<string> {
    const lines = [`Original question: ${question}\n`];
    for (let i = 0; i < steps.length; i++) {
      const step = steps[i]!;
      const result = results[i]!;
      const status = result.success ? "✓" : "✗";
      lines.push(`Step ${step.step_number} [${status}]: ${step.description}`);
      if (result.success && result.data) {
        lines.push(`Result: ${JSON.stringify(result.data)}`);
      } else if (result.error) {
        lines.push(`Error: ${result.error}`);
      }
      lines.push("");
    }

    const response = await this.client.chat.completions.create({
      model: this.model,
      messages: [
        { role: "system", content: SYNTHESIZER_SYSTEM },
        { role: "user",   content: lines.join("\n") },
      ],
      temperature: 0.5,
    });
    return response.choices[0]?.message.content ?? "";
  }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const agent = new PlanAndExecuteAgent();
  const query = "Compare Apple and Microsoft stock performance this week";

  console.log("\n" + "=".repeat(60));
  console.log(`Query: ${query}`);
  console.log("=".repeat(60));

  const output = await agent.run(query);

  console.log("\n--- Plan ---");
  for (const step of output.plan) {
    const deps = step.depends_on.length > 0 ? ` (depends on ${step.depends_on})` : "";
    const tool = step.tool_name ? ` [${step.tool_name}]` : " [reasoning]";
    console.log(`  ${step.step_number}. ${step.description}${tool}${deps}`);
  }

  console.log("\n--- Results ---");
  for (const result of output.results) {
    const status = result.success ? "✓" : "✗";
    console.log(`  Step ${result.step_number} [${status}] (${result.duration_ms}ms)`);
    if (result.success && result.data) {
      console.log(`    ${JSON.stringify(result.data)}`);
    } else if (result.error) {
      console.log(`    ERROR: ${result.error}`);
    }
  }

  console.log("\n--- Final Answer ---");
  console.log(output.answer);
}

main().catch(console.error);
